File: demos/java-feequote/src/main/java/com/feequote/jobs/Reconciliation.java
package com.feequote.jobs;

import com.feequote.legacy.Fees;
import java.util.ArrayList;
import java.util.List;

/**
 * Nightly batch reconciliation.
 *
 * <p>Recomputes the fee on every line of a settled batch the way the batch system
 * booked it and reports the lines that do not agree. The batch system still books
 * fees with the legacy truncating helper, so this job has to use the same helper to
 * reproduce historical totals. Ticket 4821 tracks moving both sides to core pricing.
 */
public final class Reconciliation {

    /**
     * One settled transfer as recorded by the batch system.
     *
     * @param transferId the transfer reference
     * @param amountCents the settled amount in cents
     * @param rateBps the rate the batch system applied
     * @param bookedFeeCents the fee the batch system actually booked
     */
    public record BatchLine(String transferId, long amountCents, int rateBps, long bookedFeeCents) {
    }

    /**
     * What a whole batch looked like once recomputed.
     *
     * @param lineCount how many lines were checked
     * @param expectedTotalCents the total the recomputation produced
     * @param bookedTotalCents the total the batch system booked
     * @param mismatchedTransferIds the transfers whose booked fee did not agree
     */
    public record ReconciliationReport(
            int lineCount,
            long expectedTotalCents,
            long bookedTotalCents,
            List<String> mismatchedTransferIds) {

        public ReconciliationReport {
            mismatchedTransferIds = List.copyOf(mismatchedTransferIds);
        }

        public boolean isBalanced() {
            return mismatchedTransferIds.isEmpty();
        }
    }

    private Reconciliation() {
    }

    /** The fee the batch system should have booked for {@code line}. */
    public static long expectedFee(BatchLine line) {
        return Fees.calculateFee(line.amountCents(), line.rateBps());
    }

    /** Compares booked fees against recomputed fees for a whole batch. */
    public static ReconciliationReport reconcile(Iterable<BatchLine> lines) {
        int lineCount = 0;
        long expectedTotal = 0;
        long bookedTotal = 0;
        List<String> mismatches = new ArrayList<>();
        for (BatchLine line : lines) {
            lineCount++;
            long expected = expectedFee(line);
            expectedTotal += expected;
            bookedTotal += line.bookedFeeCents();
            if (expected != line.bookedFeeCents()) {
                mismatches.add(line.transferId());
            }
        }
        return new ReconciliationReport(lineCount, expectedTotal, bookedTotal, mismatches);
    }
}

File: demos/python-feequote/feequote/jobs/reconciliation.py
"""Nightly batch reconciliation.

Recomputes the fee on every line of a settled batch the way the batch system
booked it and reports the lines that do not agree. The batch system still
books fees with the legacy truncating helper, so this job has to use the same
helper to reproduce historical totals. Ticket 4821 tracks moving both sides
to core pricing.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

from feequote.legacy.fees import calculate_fee


@dataclass(frozen=True)
class BatchLine:
    """One settled transfer as recorded by the batch system."""

    transfer_id: str
    amount_cents: int
    rate_bps: int
    booked_fee_cents: int


@dataclass(frozen=True)
class ReconciliationReport:
    line_count: int
    expected_total_cents: int
    booked_total_cents: int
    mismatched_transfer_ids: tuple[str, ...]

    @property
    def is_balanced(self) -> bool:
        return not self.mismatched_transfer_ids


def expected_fee(line: BatchLine) -> int:
    """The fee the batch system should have booked for ``line``."""
    return calculate_fee(line.amount_cents, line.rate_bps)


def reconcile(lines: Iterable[BatchLine]) -> ReconciliationReport:
    """Compare booked fees against recomputed fees for a whole batch."""
    line_count = 0
    expected_total = 0
    booked_total = 0
    mismatches: list[str] = []
    for line in lines:
        line_count += 1
        expected = expected_fee(line)
        expected_total += expected
        booked_total += line.booked_fee_cents
        if expected != line.booked_fee_cents:
            mismatches.append(line.transfer_id)
    return ReconciliationReport(
        line_count=line_count,
        expected_total_cents=expected_total,
        booked_total_cents=booked_total,
        mismatched_transfer_ids=tuple(mismatches),
    )
